operator/fee_recipient/controller.go

```diff
@@ -74,44 +71,61 @@ func (rc *recipientController) listenToTicker(logger *zap.Logger, slots chan pha
 	firstTimeSubmitted := false
 	for currentSlot := range slots {
 		// submit if first time or if first slot in epoch
-		slotsPerEpoch := rc.ethNetwork.SlotsPerEpoch()
-		if firstTimeSubmitted && uint64(currentSlot)%slotsPerEpoch != 0 {
+		if firstTimeSubmitted && uint64(currentSlot)%rc.ethNetwork.SlotsPerEpoch() != 0 {
 			continue
 		}
-
 		firstTimeSubmitted = true
-		// submit fee recipient
-		shares, err := rc.shareStorage.GetFilteredShares(logger, storage.ByOperatorIDAndActive(rc.operatorData.ID))
+
+		err := rc.prepareAndSubmit(logger, currentSlot)
 		if err != nil {
-			logger.Warn("could not get validators shares", zap.Error(err))
-			continue
+			logger.Warn("could not submit proposal preparations", zap.Error(err))
 		}
+	}
+}
 
-		var g multierror.Group
-		const batchSize = 500
-		var counter int32
-		for start, end := 0, 0; start <= len(shares)-1; start = end {
-			end = start + batchSize
-			if end > len(shares) {
-				end = len(shares)
-			}
-			batch := shares[start:end]
-
-			g.Go(func() error {
-				m, err := rc.toProposalPreparation(logger, batch)
-				if err != nil {
-					return errors.Wrap(err, "could not build proposal preparation")
-				}
-				atomic.AddInt32(&counter, int32(len(batch)))
-				return rc.beaconClient.SubmitProposalPreparation(m)
-			})
+func (rc *recipientController) prepareAndSubmit(logger *zap.Logger, slot phase0.Slot) error {
+	shares, err := rc.shareStorage.GetFilteredShares(logger, storage.ByOperatorIDAndActive(rc.operatorData.ID))
+	if err != nil {
+		return errors.Wrap(err, "could not get shares")
+	}
+
+	const batchSize = 500
+	var submitted int
+	for start := 0; start < len(shares); start += batchSize {
+		end := start + batchSize
+		if end > len(shares) {
+			end = len(shares)
 		}
-		if err := g.Wait().ErrorOrNil(); err != nil {
-			logger.Warn("failed to submit proposal preparation", zap.Error(err))
-		} else {
-			logger.Debug("proposal preparation submitted", zap.Int32("count", atomic.LoadInt32(&counter)))
+		batch := shares[start:end]
+
+		count, err := rc.submit(logger, batch)
```

Comment:
is it possible that some were sent successfully and we still get error? if so it's not counted in submitted

Reply:
ok it seems unlikely, resolving this.